<launch>
    <arg name="road_trajectory_topic" default="/road_trajectory"/>
    <arg name="incoming_image" default="/warped_image"/>
    <arg name="camera_info" default="/camera/camera_info"/>
    <arg name="homography_topic" default="/camera/camera_homography"/>
    <arg name="visual_detection_topic" default="/line_out"/>
    <arg name="camera_frame" default="/camera_optical"/>
    <arg name="world_frame" default="/front_axis_middle"/>
    <arg name="debug" default="false"/>
    <arg name="use_nodelet" default="false"/>
    <arg name="nodelet_manager" default=""/>
    <arg name="NM1" default="$(find drive_ros_image_recognition)/config/trained_classifierNM1.xml" />
    <arg name="NM2"  default="$(find drive_ros_image_recognition)/config/trained_classifierNM2.xml" />

    <arg if="$(arg debug)" name="launch-prefix" value="gdb -ex run --args"/>
    <arg unless="$(arg debug)" name="launch-prefix" value=""/>

    <group unless="$(arg use_nodelet)">
        <node name="speed_marking_detection" pkg="drive_ros_image_recognition" type="speed_marking_detection_node" respawn="false" output="screen" launch-prefix="$(arg launch-prefix)">
            <remap from="img_in" to="$(arg incoming_image)"/>
            <remap from="road_in" to="$(arg road_trajectory_topic)"/>
            <remap from="camera_info" to="$(arg camera_info)"/>
            <remap from="homography_in" to="$(arg homography_topic)"/>
            <remap from="line_out" to="$(arg visual_detection_topic)"/>
            <param name="world_frame" type="str" value="$(arg world_frame)"/>
            <param name="camera_frame" type="str" value="$(arg camera_frame)"/>
            <!-- todo: car command output channel, as we have not decided on the interface yet, will comment for now
            <remap from="car" to "car_out"/> -->
            <rosparam command="load" file="$(find drive_ros_image_recognition)/config/road_detection.yaml"/>
            <rosparam command="load" file="$(find drive_ros_image_recognition)/config/homography.yaml"/>
            <param name="NM1" type="string" value="$(arg NM1)"/>
            <param name="NM2" type="string" value="$(arg NM2)"/>
        </node>
    </group>

    <group if="$(arg use_nodelet)">
        <arg if="$(eval arg('nodelet_manager') == '')" name="manager_used" value="speed_marking_nodelet_manager"/>
        <arg unless="$(eval arg('nodelet_manager') == '')" name="manager_used" value="$(arg nodelet_manager)"/>

        <node if="$(eval arg('nodelet_manager') == '')" pkg="nodelet" type="nodelet" name="speed_marking_nodelet_manager"  args="manager" output="screen"/>
        <node pkg="nodelet" type="nodelet" name="speed_marking_detection_nodelet" args="load drive_ros_image_recognition/SpeedMarkingDetection $(arg manager_used)" output="screen" launch-prefix="$(arg launch-prefix)">
            <remap from="img_in" to="$(arg incoming_image)"/>
            <remap from="road_in" to="$(arg road_trajectory_topic)"/>
            <remap from="camera_info" to="$(arg camera_info)"/>
            <remap from="homography_in" to="$(arg homography_topic)"/>
            <remap from="line_out" to="$(arg visual_detection_topic)"/>
            <param name="world_frame" type="str" value="$(arg world_frame)"/>
            <param name="camera_frame" type="str" value="$(arg camera_frame)"/>
            <!-- todo: car command output channel, as we have not decided on the interface yet, will comment for now
            <remap from="car" to "car_out"/> -->
            <rosparam command="load" file="$(find drive_ros_image_recognition)/config/road_detection.yaml"/>
            <rosparam command="load" file="$(find drive_ros_image_recognition)/config/homography.yaml"/>
            <param name="NM1" type="string" value="$(arg NM1)"/>
            <param name="NM2" type="string" value="$(arg NM2)"/>
        </node>
    </group>
</launch>
